README.md
![Image](./readme.png)  

# nvr-js
![NPM](https://img.shields.io/npm/l/nvr-js)
![npm](https://img.shields.io/npm/v/nvr-js)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/marcus-j-davies/nvr-js.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/marcus-j-davies/nvr-js/context:javascript)
![npms.io (maintenance)](https://img.shields.io/npms-io/maintenance-score/nvr-js)
![Libraries.io dependency status for latest release](https://img.shields.io/librariesio/release/npm/nvr-js)

NVR JS is a simple, very lightweight and efficient CCTV NVR based on Node JS.  
it's primarily aimed for 24/7 recording and live viewing.

Under the hood it uses ffmpeg, node js and websockets, all wrapped in a web based user interface.
The NVR has an API that allows to create events and timestamp them on the 24/7 recordings.

The 24/7 recordings can be reviewed using a timeline UI where the events are also time aligned on that same timeline.

![Image](./demo.png) 

### Inspired by shinobi.video
[Shinobi](https://shinobi.video) is a fully featured, jam packed NVR, also built using Node JS.
I was using Shinobi and thought it was amazing! - however, it had sooo much to it, and was too overkill for my needs.

You can think of NVR-JS as a very slimed down version of shinobi video, built from the ground up.  
the table below, shows how slimmed down it is.

| Feature           | Shinobi | NVR JS              |
|-------------------|---------|---------------------|
| Motion Dectection | &check; |                     |
| Object Detection  | &check; |                     |
| 24/7 Recording    | &check; | &check;             |
| Event Creation    | &check; | &check; (API Only)  |
| Notifications     | &check; |                     |
| Live Streaming    | &check; | &check; (Websocket) |
| Configuration UI  | &check; | Manual Editing      |
| Mobile Support    | &check; |                     |

As you can see, NVR JS does not pack the same features as Shinobi Video, but that's the intention.  
NVR JS is designed for 24/7 recording with access to live footage, and the 24/7 recordings.

### Configuration.
The first time you start NVR JS, it will create a config file at **/%home%/nvrjs.config.js**,  
it will then terminate allowing you to start adjusting your system, before starting it up again.

```javascript
module.exports = {
	/* System Settings */
	system: {
		/* Disable Security - Know what your doing before changing this! */
		disableUISecurity: false,
		/* Username */
		username: 'admin',
		/* bcrypt password (default: admin) */
		password: '$2a$10$CnOx/6vFY2ehRDf68yqd..aLlv0UM.zeBLKnRjuU8YykCsC2Ap3iG',
		/* bcrypt API Key (default: x7Te9m38JHQq6ddv) */
		apiKey: '$2a$10$N53ci.EIQ7JCu6u1HlOjoO//W0Bmp3GrRruyK1Jysr01CQ1rDrVQK',
		/* Any random string */
		cookieKey: 'f3gi6FLhIPVV31d1TBQUPEAngrI3wAoP',
		interfacePort: 7878,
		/* location used for 24/7 recording and database generation */
		/* This should be the root of a mount point i.e a dedicated HDD for 24/7 recordings */
		storageVolume: '/Volumes/CCTV',
		/* Continuous recording settings */
		ffmpegLocation: 'ffmpeg',
		continuousSegTimeMinutes: 15,
		continuousDays: 14,
		continuousPurgeIntervalHours: 24,
		/* event throttle per sensorId */
		eventSensorIdCoolOffSeconds: 60
	},
	/* Cameras */
	cameras: {
		'66e39d21-72c4-405c-a838-05a8e8fe0742': {
			name: 'Garage',
			/* Input Source Config */
			/* The keys and values represent the ffmpeg options */
			inputConfig: {
				use_wallclock_as_timestamps: '1',
				fflags: '+igndts',
				analyzeduration: '1000000',
				probesize: '1000000',
				rtsp_transport: 'tcp'
			},
			/* Input Address */
			input: 'rtsp://user:password@ip:port/live0',
			/* Only use these if the incoming stream is not already h264/5 */
			/* Note that depending on the encoder in use, The CPU could be taxed */
			/* The below will use the HW assisted h264 encoder for RPi4 */
			postInput: {
				// videoEncoder: 'h264_v4l2m2m',
				// videoAdditional:{ffmpeg_option:'value'}
				// audioEncoder: 'aac',
				// audioAdditional:{ffmpeg_option:'value'}
			},
			/* Recording 24/7 */
			/* Disabling continuous recording, will disable the ability to create events */
			continuous: true,
			/* Live streaming config */
			/* These settings should be good enough for a low delay live stream, providing your camera produces h264 frames */
			/* streaming is achieved with websockets and MP4 fragments */
			liveConfig: {
				codecString: 'video/mp4; codecs="avc1.64001f"',
				streamConfig: {
					an: '',
					vcodec: 'copy',
					f: 'mp4',
					movflags: '+frag_keyframe+empty_moov+default_base_moof',
					reset_timestamps: '1'
				}
			}
		}
	}
};
```


### The Event API.
To create events one only needs to send the following JSON payload.

The view here, is that you create events from various sensors in your setup, this effectively acts as your motion detector
or some other key event - It's really up to you. 

```javascript
{
     "event": "Motion Detected" | "Door Opened" | "Some Other Event" | "Of Your Choice",
     "sensorId": "HUEN849",
     "timestamp": 1636194611
}
```

You **POST** this payload to the API as follows:  
http://IP:7878/api/{APIKey}/event/{camera-id}  
Example: http://IP:7878/api/x7Te9m38JHQq6ddv/event/66e39d21-72c4-405c-a838-05a8e8fe0742

### Anyway
I built this for my needs, it's very DIY but should be fairly stable (famous last words)
But if you want to use it, change it, build on it, feel free - I welcome PR's.

```
npm install nvr-js
cd ./node_modules/nvr-js
pm2 start NVRJS.js
pm2 save
```

[Change Log](./CHANGELOG.md)


CHANGELOG.md
# NVR-JS Change Log

  - 3.0.0 

    **Breaking Changes**
     - The **NVRJS_CAMERA_RECORDINGS** directory is no longer used.
     - The Storage mechanism has been updated, and is not compatible with previous recordings
        - You can continue to run this version with no issues, but previous data will not be available.
     - Event payload properties have been updated.
        - **name** -> **event**
        - **date** -> **timestamp**

    **New Features**
     - Raise an event whilst reviewing live footage
     - Each recorded segement, now contains a hash, this offers the ability to check for modifications of recorded footage (SHA256).
     - Allow post input configuration This allows to convert a stream that may not natively be h264
     - Ability to disable security, allowing 3rd party access control.

    **Changes**
     - Dependency updates.
     - SQL storage has been removed and replaced with JSON files
     - Performancee updates


  - 2.0.0

    **Breaking Changes**
     - NVR system folders have been renamed. rename them to continue with your current data.
       - system  -> NVRJS_SYSTEM  
       - cameras -> NVRJS_CAMERA_RECORDINGS
     - API Access no longer uses the UI password, it uses its own API key, as configured in the config file.  
       Add a new value named **apiKey** in the **system** section - this should be a bcript value of your chosen API key
     - Username is now requied in the login page.  
       Add a new value named **username** in the **system** section to set it - this should be plain text  
     - API is now accessed via the /api/ URI

    **Changes**
     - Dependency updates.
     - Clean/polish up the UI.
     - Re-worked ffmpeg stream pipes.
     - SQL data writes are now queued.
     - Rate Limiting is now applied to the the HTTP application.

    **New Features**
     - New API functions (URI's)
       - **/systeminfo**  
       - **/cameras**  
       - **/snapshot/:CameraID/:Width**  
       - **/geteventdata/:CameraID/:Start/:End**  

  - 1.0.2

    **Fixes**
     - Fix directory creation

  - 1.0.1

    **Fixes**
     - Correct drive space usage query.

  - 1.0.0

    **Initial Release**





NVRJS.js
const express = require('express');
const cookieparser = require('cookie-parser');
const cookie = require('cookie');
const bcrypt = require('bcrypt');
const crypto = require('crypto');
const http = require('http');
const io = require('socket.io');
const handlebars = require('handlebars');
const childprocess = require('child_process');
const MP4Frag = require('./core/MP4Frag');
const fs = require('fs');
const os = require('os');
const path = require('path');
const osu = require('node-os-utils');
const dayjs = require('dayjs');
const customParseFormat = require('dayjs/plugin/customParseFormat');
dayjs.extend(customParseFormat);
const RateLimiter = require('express-rate-limit');

/* Some static things */
const FileType = '.mp4';
const Index = {};
const SensorTimestamps = {};

/* Check and load config */
console.log(' - Checking config.');
if (!fs.existsSync(path.join(os.homedir(), 'nvrjs.config.js'))) {
	fs.copyFileSync(
		path.join(__dirname, 'nvrjs.config.example.js'),
		path.join(os.homedir(), 'nvrjs.config.js')
	);
	console.log(
		` - New config created: ${path.join(os.homedir(), 'nvrjs.config.js')}`
	);
	console.log(' - Edit config to suite and restart!');
	process.exit(0);
}
const config = require(path.join(os.homedir(), 'nvrjs.config.js'));
console.log(` - Config loaded: ${path.join(os.homedir(), 'nvrjs.config.js')}`);

/* Storage Volumes and folders */
console.log(' - Checking volumes and ffmpeg.');
if (!fs.existsSync(config.system.storageVolume)) {
	console.log(' - Storage volume does not exist');
	process.exit();
} else {
	try {
		if (
			!fs.existsSync(path.join(config.system.storageVolume, 'NVRJS_SYSTEM'))
		) {
			fs.mkdirSync(path.join(config.system.storageVolume, 'NVRJS_SYSTEM'));
		}
	} catch (e) {
		console.log('Error creating system directories.');
		console.log(e.message);
		process.exit(0);
	}
}

/* FFMPEG Check */
if (!fs.existsSync(config.system.ffmpegLocation)) {
	console.log(
		`ffmpeg not found in specifed location: ${config.system.ffmpegLocation}`
	);
	process.exit(0);
}

/* Protect web service from abuse */
const IOLimiter = RateLimiter({
	windowMs: 2000,
	max: 100
});

/* Configure WEB UI */
console.log(' - Creating express application.');
const App = new express();
App.use(IOLimiter);
App.use(express.json());
App.use(cookieparser(config.system.cookieKey));
const HTTP = new http.Server(App);

console.log(' - Compiling pages.');
const CompiledPages = {};
const Pages = {
	Dash: path.join(__dirname, 'web', 'dash.html'),
	Index: path.join(__dirname, 'web', 'index.html')
};
Object.keys(Pages).forEach((PS) => {
	CompiledPages[PS] = handlebars.compile(fs.readFileSync(Pages[PS], 'utf8'));
});

// Static
App.use('/static', express.static(path.join(__dirname, 'web', 'static')));

// UI
App.get('/', (req, res) => {
	if (
		config.system.disableUISecurity !== undefined &&
		config.system.disableUISecurity
	) {
		res.redirect('/dashboard');
	} else {
		res.type('text/html');
		res.status(200);
		res.end(CompiledPages.Index());
	}
});
App.post('/login', (req, res) => {
	const Data = req.body;
	const Password = Data.password;
	const Username = Data.username;

	if (
		bcrypt.compareSync(Password, config.system.password) &&
		config.system.username === Username
	) {
		res.cookie('Authentication', 'Success', {
			signed: true
		});
		res.status(204);
		res.end();
	} else {
		res.status(401);
		res.end();
	}
});

// Dashboard
App.get('/dashboard', CheckAuthMW, (req, res) => {
	res.type('text/html');
	res.status(200);
	res.end(CompiledPages.Dash(config));
});

// System Info (Uses shared API)
App.get('/systeminfo', CheckAuthMW, (req, res) => {
	getSystemInfo(req, res);
});

// Snapshot (Uses shared API)
App.get('/snapshot/:CameraID/:Width', CheckAuthMW, (req, res) => {
	getSnapShot(res, req.params.CameraID, req.params.Width);
});

// Get Event data (Uses shared API)
App.get('/geteventdata/:CameraID/:Start/:End', CheckAuthMW, (req, res) => {
	GetEventData(res, req.params.CameraID, req.params.Start, req.params.End);
});

/* Configure APIs */

// System Info
App.get('/api/:APIKey/systeminfo', (req, res) => {
	if (bcrypt.compareSync(req.params.APIKey, config.system.apiKey)) {
		getSystemInfo(req, res);
	} else {
		res.status(401);
		res.end();
	}
});

function getSystemInfo(req, res) {
	osu.cpu.usage().then((CPU) => {
		osu.drive.info(config.system.storageVolume).then((DISK) => {
			osu.mem.info().then((MEM) => {
				const Info = {
					CPU: CPU,
					DISK: DISK,
					MEM: MEM
				};
				res.type('application/json');
				res.status(200);
				res.end(JSON.stringify(Info));
			});
		});
	});
}

// Get Cameras
App.get('/api/:APIKey/cameras', (req, res) => {
	if (bcrypt.compareSync(req.params.APIKey, config.system.apiKey)) {
		const Cams = [];

		Object.keys(config.cameras).forEach((ID) => {
			const Cam = config.cameras[ID];
			Cams.push({
				id: ID,
				name: Cam.name,
				continuous: Cam.continuous
			});
		});

		res.type('application/json');
		res.status(200);
		res.end(JSON.stringify(Cams));
	} else {
		res.status(401);
		res.end();
	}
});

// Create Event
App.post('/event/:CameraID', CheckAuthMW, (req, res) => {
	Event(
		res,
		req.params.CameraID,
		req.body.event,
		req.body.sensorId,
		req.body.timestamp
	);
});

App.post('/api/:APIKey/event/:CameraID', (req, res) => {
	if (bcrypt.compareSync(req.params.APIKey, config.system.apiKey)) {
		Event(
			res,
			req.params.CameraID,
			req.body.event,
			req.body.sensorId,
			req.body.timestamp
		);
	} else {
		res.status(401);
		res.end();
	}
});

function Event(res, CameraID, Event, SensorID, Timestamp) {
	if (!config.cameras.hasOwnProperty(CameraID)) {
		res.status(404);
		res.end();
		return;
	}

	if (config.cameras[CameraID].continuous) {
		if (
			!SensorTimestamps.hasOwnProperty(SensorID) ||
			SensorID === 'LIVE-VIEW-EVENT'
		) {
			const Path = path.join(
				config.system.storageVolume,
				'NVRJS_SYSTEM',
				CameraID
			);

			const Last = Math.max(
				...Object.keys(Index[CameraID]).map((E) => parseInt(E))
			);
			const PHFOP = path.join(Path, Index[CameraID][Last]);
			const PHFO = ReadMetaFile(PHFOP);

			if (PHFO !== false) {
				PHFO.events.push({
					eventId: generateUUID(),
					event: Event,
					sensorId: SensorID,
					timestamp: Timestamp
				});

				const FP = path.join(Path, PHFO.segment.metaFileName);
				if (WriteMetaFile(PHFO, FP)) {
					res.status(204);
					res.end();

					SensorTimestamps[SensorID] = dayjs().unix();

					setTimeout(() => {
						delete SensorTimestamps[SensorID];
					}, 1000 * config.system.eventSensorIdCoolOffSeconds);
				} else {
					res.status(500);
					res.end();
				}
			} else {
				res.status(500);
				res.end();
			}
		} else {
			res.status(429);
			res.end();
		}
	} else {
		res.status(501);
		res.end();
	}
}

// Snapshot
App.get('/api/:APIKey/snapshot/:CameraID/:Width', (req, res) => {
	if (bcrypt.compareSync(req.params.APIKey, config.system.apiKey)) {
		getSnapShot(res, req.params.CameraID, req.params.Width);
	} else {
		res.status(401);
		res.end();
	}
});

function getSnapShot(Res, CameraID, Width) {
	const CommandArgs = [];
	const Cam = config.cameras[CameraID];

	Object.keys(Cam.inputConfig).forEach((inputConfigKey) => {
		CommandArgs.push('-' + inputConfigKey);
		if (Cam.inputConfig[inputConfigKey].length > 0) {
			CommandArgs.push(Cam.inputConfig[inputConfigKey]);
		}
	});

	CommandArgs.push('-i');
	CommandArgs.push(Cam.input);
	CommandArgs.push('-vf');
	CommandArgs.push('scale=' + Width + ':-1');
	CommandArgs.push('-vframes');
	CommandArgs.push('1');
	CommandArgs.push('-f');
	CommandArgs.push('image2');
	CommandArgs.push('-');

	const Process = childprocess.spawn(
		config.system.ffmpegLocation,
		CommandArgs,
		{ env: process.env, stderr: 'ignore' }
	);

	let imageBuffer = Buffer.alloc(0);

	Process.stdout.on('data', function (data) {
		imageBuffer = Buffer.concat([imageBuffer, data]);
	});

	Process.on('exit', (Code, Signal) => {
		const _Error = FFMPEGExitDueToError(Code, Signal);
		if (!_Error) {
			Res.type('image/jpeg');
			Res.status(200);
			Res.end(Buffer.from(imageBuffer, 'binary'));
		} else {
			Res.status(500);
			Res.end();
		}
	});
}

// Get Event Data
App.get('/api/:APIKey/geteventdata/:CameraID/:Start/:End', (req, res) => {
	if (bcrypt.compareSync(req.params.APIKey, config.system.apiKey)) {
		GetEventData(res, req.params.CameraID, req.params.Start, req.params.End);
	} else {
		res.status(401);
		res.end();
	}
});

function GetEventData(res, CameraID, Start, End) {
	const Data = {
		segments: []
	};

	const Segments = Object.keys(Index[CameraID]).filter(
		(K) => parseInt(K) >= parseInt(Start) && parseInt(K) <= parseInt(End)
	);

	Segments.forEach((K) => {
		const FilePath = path.join(
			config.system.storageVolume,
			'NVRJS_SYSTEM',
			CameraID,
			Index[CameraID][K]
		);

		const PL = ReadMetaFile(FilePath);
		if (PL !== false) {
			Data.segments.push(PL);
		}
	});

	res.type('application/json');
	res.status(200);
	res.end(JSON.stringify(Data));
}

/* FFMPEG Exit code check */
function FFMPEGExitDueToError(Code, Signal) {
	if (Code == null && Signal === 'SIGKILL') {
		return false;
	}
	if (Code === 255 && Signal == null) {
		return false;
	}
	if (Code > 0 && Code < 255 && Signal == null) {
		return true;
	}
}

/* Start up cameras */
const Cameras = Object.keys(config.cameras);
Cameras.forEach((cameraID) => {
	const Cam = config.cameras[cameraID];
	InitCamera(Cam, cameraID);
});

/* Start Purge timer */
console.log(' - Strting purge timer.');
setInterval(
	purgeContinuous,
	60000 * 60 * config.system.continuousPurgeIntervalHours
);

/* Create PlaceHolder Metafile */
function CreatePlaceHolderMeta(CameraID) {
	const Path = path.join(config.system.storageVolume, 'NVRJS_SYSTEM', CameraID);
	const Start = dayjs().unix();

	const Meta = {
		segment: {
			metaFileName: `${Start}_placeholder.json`,
			cameraId: CameraID,
			startTime: Start,
			fileName: undefined,
			endTime: 0,
			checksum: 0,
			segmentId: generateUUID()
		},
		events: []
	};

	const FP = path.join(Path, Meta.segment.metaFileName);
	if (WriteMetaFile(Meta, FP)) {
		Index[CameraID][Start] = Meta.segment.metaFileName;
		return true;
	} else {
		return false;
	}
}

/* Create Metafile */
function CreateMeta(CameraID, fileName) {
	fileName = fileName.trim().replace(/\n/g, '');
	const Path = path.join(config.system.storageVolume, 'NVRJS_SYSTEM', CameraID);

	const Start = parseInt(fileName.split('.')[0]);
	const End = dayjs().unix();

	const Meta = {
		segment: {
			metaFileName: fileName.replace(FileType, '.json'),
			cameraId: CameraID,
			fileName: fileName,
			startTime: Start,
			endTime: End,
			checksum: 0,
			segmentId: generateUUID()
		},
		events: []
	};

	const fileBuffer = fs.readFileSync(path.join(Path, fileName));
	const HashSum = crypto.createHash('sha256');
	HashSum.update(fileBuffer);

	const Hex = HashSum.digest('hex');
	Meta.segment.checksum = `sha256:${Hex}`;

	const Last = Math.max(
		...Object.keys(Index[CameraID]).map((E) => parseInt(E))
	);

	const PHFOP = path.join(Path, Index[CameraID][Last]);
	const PHFO = ReadMetaFile(PHFOP);

	if (PHFO !== false) {
		Meta.events = PHFO.events;
		if (WriteMetaFile(Meta, path.join(Path, Meta.segment.metaFileName))) {
			delete Index[CameraID][Last];
			fs.unlinkSync(PHFOP);
			Index[CameraID][Start] = Meta.segment.metaFileName;
			CreatePlaceHolderMeta(CameraID);
		}
	}
}

/* Camera Initer */
function InitCamera(Cam, cameraID) {
	console.log(` - Configuring camera: ${Cam.name}`);

	const Path = path.join(config.system.storageVolume, 'NVRJS_SYSTEM', cameraID);
	if (!fs.existsSync(Path)) {
		fs.mkdirSync(Path);
	}

	Index[cameraID] = {};

	const SegMetaFiles = fs
		.readdirSync(Path)
		.filter(
			(File) => path.extname(File) === '.json' && !File.includes('_placeholder')
		);

	SegMetaFiles.forEach((MF) => {
		const MD = ReadMetaFile(path.join(Path, MF));
		if (MD !== false) {
			Index[cameraID][MD.segment.startTime] = MD.segment.metaFileName;
		}
	});

	const CommandArgs = [];

	Object.keys(Cam.inputConfig).forEach((inputConfigKey) => {
		if (inputConfigKey !== 'i') {
			CommandArgs.push('-' + inputConfigKey);
			if (Cam.inputConfig[inputConfigKey].length > 0) {
				CommandArgs.push(Cam.inputConfig[inputConfigKey]);
			}
		}
	});

	CommandArgs.push('-i');
	CommandArgs.push(Cam.input);

	App.use(
		'/segments/' + cameraID,
		CheckAuthMW,
		express.static(
			path.join(config.system.storageVolume, 'NVRJS_SYSTEM', cameraID),
			{ acceptRanges: true }
		)
	);

	if (Cam.continuous !== undefined && Cam.continuous) {
		let CV = 'copy';
		let CA = 'copy';

		if (Cam.postInput !== undefined) {
			if (Cam.postInput.videoEncoder !== undefined) {
				CV = Cam.postInput.videoEncoder;
			}
			if (Cam.postInput.audioEncoder !== undefined) {
				CA = Cam.postInput.audioEncoder;
			}
		}
		CommandArgs.push('-c:v');
		CommandArgs.push(CV);

		if (
			Cam.postInput !== undefined &&
			Cam.postInput.videoAdditional !== undefined
		) {
			Object.keys(Cam.postInput.videoAdditional).forEach((EK) => {
				CommandArgs.push(`-${EK}`);
				CommandArgs.push(Cam.postInput.videoAdditional[EK]);
			});
		}

		CommandArgs.push('-c:a');
		CommandArgs.push(CA);

		if (
			Cam.postInput !== undefined &&
			Cam.postInput.audioAdditional !== undefined
		) {
			Object.keys(Cam.postInput.audioAdditional).forEach((EK) => {
				CommandArgs.push(`-${EK}`);
				CommandArgs.push(Cam.postInput.audioAdditional[EK]);
			});
		}

		CommandArgs.push('-f');
		CommandArgs.push('segment');
		CommandArgs.push('-segment_atclocktime');
		CommandArgs.push('1');
		CommandArgs.push('-reset_timestamps');
		CommandArgs.push('1');
		CommandArgs.push('-strftime');
		CommandArgs.push('1');
		CommandArgs.push('-segment_list');
		CommandArgs.push('pipe:4');
		CommandArgs.push('-segment_time');
		CommandArgs.push(60 * config.system.continuousSegTimeMinutes);
		CommandArgs.push('-movflags');
		CommandArgs.push('+faststart');
		CommandArgs.push(path.join(Path, `%s${FileType}`));
	}

	Object.keys(Cam.liveConfig.streamConfig).forEach((streamingConfigKey) => {
		CommandArgs.push('-' + streamingConfigKey);
		if (Cam.liveConfig.streamConfig[streamingConfigKey].length > 0) {
			CommandArgs.push(Cam.liveConfig.streamConfig[streamingConfigKey]);
		}
	});

	CommandArgs.push('-metadata');
	CommandArgs.push('title="NVR JS Stream"');
	CommandArgs.push('pipe:3');

	const Options = {
		detached: true,
		stdio: ['ignore', 'ignore', 'ignore', 'pipe', 'pipe']
	};
	const respawn = (Spawned) => {
		const MP4F = new MP4Frag();

		const IOptions = {
			path: '/streams/' + cameraID
		};
		const Socket = io(HTTP, IOptions);
		Socket.on('connection', (ClientSocket) => {
			if (CheckAuthMW(ClientSocket)) {
				ClientSocket.emit('segment', MP4F.initialization);
			}
		});

		MP4F.on('segment', (data) => {
			Socket.sockets.sockets.forEach((ClientSocket) => {
				ClientSocket.emit('segment', data);
			});
		});

		Spawned.on('close', () => {
			console.log(
				` - Camera: ${Cam.name}  was terminated, respawning after 10 seconds...`
			);
			Spawned.kill();
			MP4F.destroy();
			setTimeout(() => {
				respawn(
					childprocess.spawn(config.system.ffmpegLocation, CommandArgs, Options)
				);
			}, 10000);
		});

		Spawned.stdio[3].on('data', (data) => {
			MP4F.write(data, 'binary');
		});
		Spawned.stdio[4].on('data', (FN) => {
			CreateMeta(cameraID, FN.toString());
		});

		return Spawned;
	};

	const Process = respawn(
		childprocess.spawn(config.system.ffmpegLocation, CommandArgs, Options)
	);

	if (!CreatePlaceHolderMeta(cameraID)) {
		Process.removeAllListeners('close');
		Process.kill();
		console.log(
			`   - Could not kick start segment management for camera: ${Cam.name}`
		);
		console.log(`   - Camera will not be started.`);
	} else {
		console.log(`   - Camera: ${Cam.name} started.`);
	}
}

/* Gen ID */
function generateUUID() {
	var d = new Date().getTime();
	var d2 =
		(typeof performance !== 'undefined' &&
			performance.now &&
			performance.now() * 1000) ||
		0;
	return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
		var r = Math.random() * 16;
		if (d > 0) {
			r = (d + r) % 16 | 0;
			d = Math.floor(d / 16);
		} else {
			r = (d2 + r) % 16 | 0;
			d2 = Math.floor(d2 / 16);
		}
		return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
	});
}

/* Logged in check */
function CheckAuthMW(req, res, next) {
	const NonInteractive = res === undefined && next === undefined;

	if (
		config.system.disableUISecurity !== undefined &&
		config.system.disableUISecurity
	) {
		if (NonInteractive) {
			return true;
		} else {
			next();
		}
	} else {
		if (NonInteractive) {
			if (req.handshake.headers.cookie !== undefined) {
				const CS = cookie.parse(req.handshake.headers.cookie);
				const Signed = cookieparser.signedCookies(CS, config.system.cookieKey);
				if (
					Signed.Authentication === undefined ||
					Signed.Authentication !== 'Success'
				) {
					req.disconnect();
					return false;
				} else {
					return true;
				}
			} else {
				req.disconnect();
				return false;
			}
		} else {
			if (
				req.signedCookies.Authentication === undefined ||
				req.signedCookies.Authentication !== 'Success'
			) {
				res.status(401);
				res.end();
			} else {
				next();
			}
		}
	}
}

/* Purge */
function purgeContinuous() {
	console.log(' - Purging data.');

	const Date = dayjs().subtract(config.system.continuousDays, 'day').unix();

	Object.keys(Index).forEach((K) => {
		const Files = Object.keys(Index[K]).filter((TSK) => parseInt(TSK) <= Date);
		Files.forEach((F) => {
			const Path = path.join(config.system.storageVolume, 'NVRJS_SYSTEM', K);
			delete Index[K][F]; // Index Entry
			fs.unlinkSync(path.join(Path, `${F}.json`)); // Metafile
			fs.unlinkSync(path.join(Path, `${F}${FileType}`)); // footage
		});
	});
}

/* Write Operation */
function WriteMetaFile(OBJ, FilePath) {
	try {
		fs.writeFileSync(FilePath, JSON.stringify(OBJ));
		return true;
	} catch (Err) {
		return false;
	}
}

/* Read Operation */
function ReadMetaFile(FilePath) {
	try {
		const Result = fs.readFileSync(FilePath, 'utf8');
		const OBJ = JSON.parse(Result);
		return OBJ;
	} catch (Err) {
		return false;
	}
}

HTTP.listen(config.system.interfacePort);
console.log(' - NVR JS is Ready!');

/* Why not */
setTimeout(purgeContinuous, 1000);


nvrjs.config.example.js
module.exports = {
	/* System Settings */
	system: {
		/* Username */
		username: "admin",
		/* bcrypt password (default: admin) */
		password: '$2a$10$CnOx/6vFY2ehRDf68yqd..aLlv0UM.zeBLKnRjuU8YykCsC2Ap3iG',
		/* bcrypt API Key (default: x7Te9m38JHQq6ddv) */
		apiKey: '$2a$10$N53ci.EIQ7JCu6u1HlOjoO//W0Bmp3GrRruyK1Jysr01CQ1rDrVQK',
		/* Any random string */
		cookieKey: 'f3gi6FLhIPVV31d1TBQUPEAngrI3wAoP',
		interfacePort: 7878,
		/* location used for 24/7 recording and database generation */
		/* This should be the root of a mount point i.e a dedicated HDD for 24/7 recordings */
		storageVolume: '/Volumes/CCTV',
		/* Continuous recording settings */
		ffmpegLocation: 'ffmpeg',
		continuousSegTimeMinutes: 15,
		continuousDays: 14,
		continuousPurgeIntervalHours: 24,
		/* event throttle per sensorId */
		eventSensorIdCoolOffSeconds: 60
	},
	/* Cameras */
	cameras: {
		'66e39d21-72c4-405c-a838-05a8e8fe0742': {
			name: 'Garage',
			/* Input Source Config */
			/* The keys and values represent the ffmpeg options */
			inputConfig: {
				use_wallclock_as_timestamps: '1',
				fflags: '+igndts',
				analyzeduration: '1000000',
				probesize: '1000000',
				rtsp_transport: 'tcp'
			},
			/* Input Address */
			input: 'rtsp://user:password@ip:port/live0',
			/* Recording 24/7 */
			/* Disabling continuous recording, will disable the ability to create events */
			continuous: true,
			/* Live streaming config */
			/* These settings should be good enough for a low delay live stream, providing your camera produces h264 frames */
			/* streaming is achieved with websockets and MP4 fragments */
			liveConfig: {
				codecString: 'video/mp4; codecs="avc1.64001f"',
				streamConfig: {
					an: '',
					vcodec: 'copy',
					f: 'mp4',
					movflags: '+frag_keyframe+empty_moov+default_base_moof',
					reset_timestamps: '1'
				}
			}
		}
	}
};
